src/v/archival/ntp_archiver_service.cc

```diff
@@ -1581,18 +1522,87 @@ ntp_archiver::schedule_single_upload(const upload_context& upload_ctx) {
         .max_timestamp = upload.max_timestamp,
         .delta_offset = delta,
         .ntp_revision = _rev,
-        .archiver_term = upload_ctx.archiver_term,
+        .archiver_term = archiver_term,
         .segment_term = upload.term,
         .delta_offset_end = delta_offset_next,
         .sname_format = cloud_storage::segment_name_format::v3,
         .metadata_size_hint = tx_size,
       },
       .name = upload.exposed_name, .delta = offset - base,
       .stop = ss::stop_iteration::no,
-      .upload_kind = upload_ctx.upload_kind,
+      .upload_kind = upload_kind,
     };
 }
 
+ss::future<ntp_archiver::scheduled_upload>
+ntp_archiver::schedule_single_upload(const upload_context& upload_ctx) {
+    auto start_upload_offset = upload_ctx.start_offset;
+    auto last_stable_offset = upload_ctx.last_offset;
+
+    auto log = _parent.log();
+
+    candidate_creation_result candidate_result;
+
+    switch (upload_ctx.upload_kind) {
+    case segment_upload_kind::non_compacted:
+        candidate_result = co_await _policy.get_next_candidate(
+          start_upload_offset,
+          last_stable_offset,
+          log,
+          *_parent.get_offset_translator_state(),
+          _conf->segment_upload_timeout);
+        break;
+    case segment_upload_kind::compacted:
+        const auto& m = manifest();
+        candidate_result = co_await _policy.get_next_compacted_segment(
+          start_upload_offset, log, m, _conf->segment_upload_timeout);
+        break;
+    }
+
+    co_return co_await ss::visit(
+      candidate_result,
+      [](std::monostate) -> ss::future<scheduled_upload> {
+          vassert(false, "Unexpected default upload candidate creation result");
+      },
+      [this, upload_ctx](upload_candidate_with_locks& candidate_with_locks) {
```

Comment:
any reason `upload_ctx` is captured by copy here instead of by ref?

Reply:
it was an oversight, fixed